query: support rawConfig queries

stefanhengl opened the pull request:
This change removes visibility queries in favor of `query.RawConfig` queries.
```
// query all private forks
q = query.RcOnlyForks|query.RcOnlyPrivate
```

At query time, the mask is compared to corresponding mask in 
indexData.rawConfigMasks.

A repository's mask is computed based on its RawConfig when we read the
index file.

We want to support these types of queries to avoid db lookups on
the critical path in Sourcegraph.

stefanhengl committed:
query: support rawConfig queries

This change removes visibility queries in favor of query.RawConfig queries.
query.RawConfig comes with a constructor that takes an optional mask of
flags (Public, Private, Forks, NoForks, Archived, NoArchived).

For example, to query all private forks:
```
q = query.NewRawConfig(query.Forks|query.Private)

```

The default flag is `Public|NoForks|NoArchived`.

At query time, the mask is compared to a similar mask in the
repository's meta data.

A repository's mask is created based on RawConfig when we read the
IndexFile.

rmmh commented on read_test.go:
```diff
@@ -213,3 +214,43 @@ func TestReadSearch(t *testing.T) {
+		rawConfig map[string]string
+		want      string
+	}{
+		{
```
test the empty map?

rmmh commented on api.go:
```diff
@@ -244,6 +244,9 @@ type Repository struct {
 	RawConfig map[string]string
 
+	// Encoded RawConfig.
+	RawConfigEncoded uint8
```
Should this be part of the public API? It seems to be more of an internal implementation detail.

keegancsmith replied:
I agree, it doesn't feel like it should be part of the public API. It feels like something that should live in indexdata. Like we do for branches.

stefanhengl replied:
moved to indexData

stefanhengl commented on read.go:
```diff
@@ -155,6 +155,10 @@ func (r *reader) readIndexData(toc *indexTOC) (*indexData, error) {
 	d.repoMetaData = []Repository{*repo}
 
+	for _, md := range d.repoMetaData {
+		md.RawConfigEncoded = EncodeRawConfig(md.RawConfig)
```
Good catch!

stefanhengl committed:
address PR comments, part 1

keegancsmith commented on matchtree.go:
```diff
@@ -957,18 +957,12 @@ func (d *indexData) newMatchTree(q query.Q) (matchTree, error) {
 			numDocs: d.numDocs(),
 			predicate: func(docID uint32) bool {
-				return repoVisible[d.repos[docID]]
+				return s.Encoded&d.repoMetaData[d.repos[docID]].RawConfigEncoded == s.Encoded
```
we could get rid of translating into a bitmask and instead calculate the predicate once per repo rather than once per doc.

stefanhengl replied:
I moved the encoded RawConfig to indexData and consequently removed it from the api. It feels much better now IMO.

I kept everything on document level because otherwise we would have to introduce the assumption that all documents belonging to a repo are stored together in a compound shard. In the future, we might want to tombstone and append single documents which is straightforward to do if we keep everything document based. WDYT?

keegancsmith commented on query/parse.go:
```diff
@@ -383,26 +375,23 @@ var tokNames = map[int]string{
-	"t:":          tokType,
-	"type:":       tokType,
-	"v:":          tokVis,
-	"visibility:": tokVis,
```
so you are removing query language support? That sounds good to me.

keegancsmith commented on shards/shards_test.go:
```diff
@@ -513,76 +513,82 @@ func BenchmarkShardedSearch(b *testing.B) {
-		if public {
-			r.RawConfig = map[string]string{"public": "1"}
-		}
+		r.RawConfigEncoded = zoekt.EncodeRawConfig(rawConfig)
```
this is somewhat of a sign to me that this is wrong? IE we shouldn't need to specify both RawConfigEncoded and RawConfig when creating a Repository. It should be created in the indexData from Repository.

keegancsmith reviewed (commented):
The RawConfigEncoded in the API seems weird. Especially since you needed to specify it to get your tests to work.

Is there a way to just synthesize this into indexdata at shard load time? Alternatively, can we just directly query RawConfig in a performant way at query time?

stefanhengl committed:
fix tests

stefanhengl committed:
remove RawConfigEncoded from api

this moves RawConfigEncoded to []rawConfigMaks in indexData as suggested
in the PR comments.

stefanhengl committed:
align flag names with Sourcegraph

keegancsmith commented on query/query.go:
```diff
@@ -30,12 +30,52 @@ var _ = log.Println
+// RawConfig filters for repositories based on their encoded RawConfig map.
+// Always use the constructor NewRawConfig instead of creating RawConfig
+// manually.
+type RawConfig struct {
```
alternatively just do something like

```
type RawConfig uint8

func (r RawConfig) String() string { ... }

const (
  Public RawConfig = 1
  ...
)
```

Then to use the API you can avoid the constructor and just directly use the RawConfig type. eg `query.NewAnd(q, query.Public | query.Archived)`

stefanhengl replied:
Yep, I agree. I wanted the constructor to offer the user a reasonable default, but I think the simpliciy of using 'type RawConfig uint64' outweighs the benefits of the constructor.

keegancsmith commented on query/query.go:
```diff
@@ -30,12 +30,52 @@ var _ = log.Println
+	OnlyForks          = 1 << 2
+	NoForks            = 2 << 2
+	OnlyArchived       = 1 << 4
+	NoArchived         = 2 << 4
```
The flags probably need a common prefix so we know they are related. For example `regexp/syntax` prefixes all constants in its bit flag with `Op`. I think that will lead to more understandable namespacing when using auto-complete/godoc/etc.

stefanhengl replied:
I picked 'Rc'. I saw that the 'os' package uses 'O_<all caps>' for the flags, but I think 'Rc' is more conventional.

stefanhengl committed:
add prefix "Rc"

stefanhengl committed:
use type RawConfig uint8 instead

stefanhengl committed:
uint8 -> uint64

stefanhengl committed:
update docstrings
